File: examples/python/normal_estimation.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-

import numpy as np
import time
import open3d as o3d
import misc3d as m3d

vis = o3d.visualization.Visualizer()
vis.create_window("Estimate normals", 1920, 1200)

depth = o3d.io.read_image('../data/indoor/depth/depth_0.png')
color = o3d.io.read_image('../data/indoor/color/color_0.png')

rgbd = o3d.geometry.RGBDImage.create_from_color_and_depth(
    color, depth, convert_rgb_to_intensity=False)

pinhole_camera_intrinsic = o3d.camera.PinholeCameraIntrinsic(
    848, 480, 598.7568, 598.7568, 430.3443, 250.244)
pcd = o3d.geometry.PointCloud.create_from_rgbd_image(
    rgbd, pinhole_camera_intrinsic, project_valid_depth_only=False)

t0 = time.time()
m3d.common.estimate_normals(pcd, (848, 480), 3)
print('time cost: {}'.format(time.time() - t0))

m3d.vis.draw_pose(vis, size=0.1)
m3d.vis.draw_geometry3d(vis, pcd)
vis.run()

File: examples/cpp/normal_estimation.cpp
#include <iostream>
#include <memory>

#include <misc3d/common/normal_estimation.h>
#include <misc3d/utils.h>
#include <misc3d/vis/vis_utils.h>
#include <open3d/camera/PinholeCameraIntrinsic.h>
#include <open3d/geometry/Image.h>
#include <open3d/geometry/PointCloud.h>
#include <open3d/geometry/RGBDImage.h>
#include <open3d/io/ImageIO.h>

int main(int argc, char *argv[]) {
    bool ret;
    open3d::geometry::Image depth, color;
    ret = open3d::io::ReadImage("../examples/data/indoor/depth/depth_0.png",
                                depth);
    ret = open3d::io::ReadImage("../examples/data/indoor/color/color_0.png",
                                color);

    auto rgbd = open3d::geometry::RGBDImage::CreateFromColorAndDepth(
        color, depth, 1000, 3.0, false);

    open3d::camera::PinholeCameraIntrinsic intrinsic(
        848, 480, 598.7568, 598.7568, 430.3443, 250.244);

    auto pcd = open3d::geometry::PointCloud::CreateFromRGBDImage(
        *rgbd, intrinsic, Eigen::Matrix4d::Identity(), false);

    misc3d::Timer timer;
    timer.Start();
    misc3d::common::EstimateNormalsFromMap(pcd, {848, 480}, 3);
    std::cout << "Time cost: " << timer.Stop() << std::endl;

    auto vis = std::make_shared<open3d::visualization::Visualizer>();
    vis->CreateVisualizerWindow("Estimate normals form map", 1920, 1200);
    misc3d::vis::DrawPose(vis, Eigen::Matrix4d::Identity(), 0.1);
    misc3d::vis::DrawGeometry3D(vis, pcd);
    vis->Run();

    return 0;
}